### PR TITLE
WIP: replacing C style includes with their correct C++ equivalents

### PR DESCRIPTION
**Pull request by slowriot**
...resolves some issues building the Vulkan samples under g++ 5.4.

Fixes the following error:
```
[ 26%] Building CXX object glslang/CMakeFiles/glslang.dir/MachineIndependent/intermOut.cpp.o
In file included from /home/slowriot/code/libraries/VulkanSDK/VulkanSamples/external/glslang/glslang/MachineIndependent/../Include/InfoSink.h:39:0,
                 from /home/slowriot/code/libraries/VulkanSDK/VulkanSamples/external/glslang/glslang/MachineIndependent/intermOut.cpp:38:
/home/slowriot/code/libraries/VulkanSDK/VulkanSamples/external/glslang/glslang/MachineIndependent/intermOut.cpp: In function ‘bool {anonymous}::is_positive_infinity(double)’:
/home/slowriot/code/libraries/VulkanSDK/VulkanSamples/external/glslang/glslang/MachineIndependent/intermOut.cpp:52:15: error: expected unqualified-id before ‘(’ token
   return std::isinf(x) && (x >= 0);
               ^
glslang/CMakeFiles/glslang.dir/build.make:374: recipe for target 'glslang/CMakeFiles/glslang.dir/MachineIndependent/intermOut.cpp.o' failed
```

### CODE REVIEW

**Comment by johnkslang** [comment]
Would like some data as to the effect this has on all the other build environments that are currently working.

File: glslang/MachineIndependent/glslang_tab.cpp [hunk 1]
```diff
@@ -2526,7 +2526,7 @@ while (0)
 #if YYDEBUG
 
 # ifndef YYFPRINTF
-#  include <stdio.h> /* INFRINGES ON USER NAME SPACE */
+#  include <cstdio> /* INFRINGES ON USER NAME SPACE */
```
Comment (dneto0) [inline glslang/MachineIndependent/glslang_tab.cpp]: I think this is a generated file, so this would be an impermanent change.

File: glslang/OSDependent/Unix/ossource.cpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ static void DetachThreadLinux(void *)
 void OS_CleanupThreadData(void)
 {
 #ifdef __ANDROID__
-	DetachThreadLinux(NULL);
+  DetachThread();
```
Comment (dneto0) [inline glslang/OSDependent/Unix/ossource.cpp]: This change is unrelated to the other changes.  I recommend not doing this in this CL.

File: glslang/OSDependent/Windows/ossource.cpp [hunk 3]
```diff
@@ -134,7 +134,7 @@ unsigned int __stdcall EnterGenericThread (void* entry)
 
 void* OS_CreateThread(TThreadEntrypoint entry)
 {
-    return (void*)_beginthreadex(0, 0, EnterGenericThread, (void*)entry, 0, 0);
+    return (void*)_beginthreadex(0, 0, EnterGenericThread, entry, 0, 0);
```
Comment (dneto0) [inline glslang/OSDependent/Windows/ossource.cpp]: This change is unrelated to the header changes.  I recommend not doing this in this CL.

File: glslang/OSDependent/Unix/ossource.cpp (hunk 2 above)
Comment (dneto0) [inline glslang/OSDependent/Unix/ossource.cpp]: In fact, the comment at the definition of DetachThreadLinux explains that it's needed because of pthread_cleanup_push demands a different function signature.

**Comment by dneto0** [comment]
I've tested this change successfully as follows:
 - Linux x86-64 with GCC 4.8.4
 - OSX with the XCode 7.3.1 compilers (Apple LLVM 7.3.0 (clang-703.0.31))
 - Cross compile to Android x86 and Android armeabi-v7a using the Android NDK r10e
 - Built the NDK from latest sources (and including this glslang commit), which exercises cross compilation to all currently supported ABIs.

The main hole in my testing is Windows.   :-)

So overall the change is good.  But I've commented on two changes in this MR that are unrelated to the header issue, and at least one of them might introduce a failure.

**Comment by johnkslang** [comment]
I can do the Windows testing, but would like to confine it to the #include changes, if that works.

@slowriot Can we see a subsetted PR for testing and pulling? Thanks.

File: glslang/MachineIndependent/intermOut.cpp [hunk 4]
```diff
@@ -38,11 +38,11 @@
 #include "../Include/InfoSink.h"
 
 #ifdef _MSC_VER
-#include <float.h>
+#include <cfloat>
 #elif defined __ANDROID__ || defined __linux__ || __MINGW32__ || __MINGW64__
 #include <cmath>
 #else
-#include <math.h>
+#include <cmath>
```
Comment (baldurk) [inline glslang/MachineIndependent/intermOut.cpp]: This seems strange to me since there's an `#elif` above for including `<cmath>`. As it stands this now becomes redundant which I don't think is the intention. I'm wondering why this even has to change - if you're using g++ shouldn't it be on one of those platforms in the `#elif`?

I seem to remember there was a bit of back and forth on this bit of code in other issues so it'd be good to be sure this change won't break some other configuration/compiler. Seems like the easier fix would be to find whichever platform is breaking and add it to the `#elif`, no?

**Comment by baldurk** [comment]
FWIW, compiles on VS2010 for me (which I figured might one of the lowest spec compilers & more likely than others not to support modern header syntax).

File: glslang/MachineIndependent/intermOut.cpp (hunk 4 above)
Comment (johnkslang) [inline glslang/MachineIndependent/intermOut.cpp]: I suspect the "math.h" was relegated as "platforms no one knows about", and is just hypothetical, and that the real back and forth here was getting the Windows/Non-Windows distinction correct.

I'm making a different PR to address this, and a broader stab at what this PR was attempting.

**Comment by johnkslang** [comment]
Since answers and new PR have not shown up, and transferring this discussion to PR #366.
